cpu/samd21/periph/adc.c

```diff
@@ -0,0 +1,361 @@
+        resolution = ADC_0_RES_8BIT;
+        break;
+    case ADC_0_RES_10BIT:
+        resolution = ADC_0_RES_10BIT;
+        break;
+    case ADC_0_RES_12BIT:
+        resolution = ADC_0_RES_12BIT;
+        break;
+    case ADC_0_RES_16BIT:
+        divideResult = ADC_0_DIV_RES_DISABLE;
+        switch(ADC_0_ACCUM_DEFAULT)
+        {
+            case ADC_0_ACCUM_512:
+                accumulate = ADC_0_ACCUM_512;
+                break;
+            case ADC_0_ACCUM_1024:
+                accumulate = ADC_0_ACCUM_1024;
+                break;
+            default:
+                accumulate = ADC_0_ACCUM_256;
+                break;
+        }        
+        resolution = ADC_0_RES_16BIT;
+        break;
+    default:
+        DEBUG("Invalid precision");        
+        return -1;
+    }
+
+    /* Set the accumlation and devide result */
+    adc->AVGCTRL.reg = ADC_AVGCTRL_ADJRES(divideResult) | accumulate;
+
+    /* Set Sample length */
+    adc->SAMPCTRL.reg = (ADC_0_SAMPLE_LENGTH << ADC_SAMPCTRL_SAMPLEN_Pos);
+    while(adc_syncing(adc));
+    
+    /* If external vref. Pin setup */
+    if(ADC_0_REF_DEFAULT == ADC_0_REF_EXT_B)
+    {
+        ADC_0_PORT.DIRCLR.reg = (1 << ADC_0_REF_DEFAULT);
+        ADC_0_PORT.PINCFG[ADC_0_REF_DEFAULT].bit.INEN = true;
+        ADC_0_PORT.PINCFG[ADC_0_REF_DEFAULT].bit.PULLEN = false;
+    }
+
+    /* Configure CTRLB Register HERE IS THE RESOLUTION SET!*/
+    adc->CTRLB.reg = 
+        ADC_0_PRESCALER |
+        resolution | 
+        (ADC_0_CORRECTION_EN << ADC_CTRLB_CORREN_Pos) | 
+        (ADC_0_FREE_RUNNING << ADC_CTRLB_FREERUN_Pos) | 
+        (ADC_0_LEFT_ADJUST << ADC_CTRLB_LEFTADJ_Pos) |
+        (ADC_0_DIFFERENTIAL_MODE << ADC_CTRLB_DIFFMODE_Pos);        
+
+
+    /* Configure Window Mode Register */
+    /*adc->WINCTRL.reg = ADC_0_WINDOW_MODE;
+    while(adc_syncing(adc));*/
+
+    /* Configure lower threshold */
+    adc->WINLT.reg = ADC_0_WINDOW_LOWER << ADC_WINLT_WINLT_Pos;
+    while(adc_syncing(adc));
+
+    /* Configure lower threshold */
+    adc->WINUT.reg = ADC_0_WINDOW_HIGHER << ADC_WINUT_WINUT_Pos;
+    while(adc_syncing(adc));
+
+    /* Configure PIN SCAN MODE & positive & Negative Input Pins */
+    adc->INPUTCTRL.reg = 
+        ADC_0_GAIN_FACTOR_DEFAULT |
+        (ADC_0_PIN_SCAN_OFFSET_START << ADC_INPUTCTRL_INPUTOFFSET_Pos) |
+        (ADC_0_PIN_SCAN_INPUT_TO_SCAN << ADC_INPUTCTRL_INPUTSCAN_Pos) |
+        ADC_0_NEG_INPUT |
+        ADC_0_POS_INPUT;
+    
+    /* Configure event action */
+    adc->EVCTRL.reg = ADC_0_EVENT_ACTION;
+
+    if (ADC_0_CORRECTION_EN)
+    {
+        if (ADC_0_GAIN_CORRECTION > ADC_GAINCORR_GAINCORR_Msk) 
+        {
+            DEBUG("Invalid gain correction\n");
+            return -1;
+        } 
+        else 
+        {
+            /* Set gain correction value */
+            adc->GAINCORR.reg = (ADC_0_GAIN_CORRECTION << ADC_GAINCORR_GAINCORR_Pos);
+        }
+        if (ADC_0_OFFSET_CORRECTION > 2047 || ADC_0_OFFSET_CORRECTION < -2048) 
```

Comment:
Using the macro 
`#define ADC_0_OFFSET_CORRECTION                ADC_OFFSETCORR_OFFSETCORR(value)`
 I had to cast `ADC_0_OFFSET_CORRECTION` to `(int)` even if the actual value is betewwn `-2048...2047`